### PR TITLE
feat: resolve latest patch for a major.minor version

### LINKED ISSUE #109: Feature Request: install latest patch version
### Feature request

Is it possible to tell the action to install the latest patch version?
e.g. latest 3.12.x increment:

```
      - uses: Azure/setup-helm@v3
        with:
          version: 'v3.12.x'
```

### PR DESCRIPTION
**Pull request by ejaifeobuks**
This pull request adds support for specifying Helm versions as major.minor (e.g., `3.14`) or with a wildcard patch (e.g., `3.14.x` or `3.14.*`) in addition to full semantic versions. It introduces logic to resolve these inputs to the latest available patch version by probing the download host. The tests have been expanded to cover these new behaviors, and error handling has been improved for version validation and network scenarios.

**Version resolution enhancements:**

* Added `isMajorMinorShaped` to detect major.minor and wildcard patch versions, and updated validation to accept these forms in `getVersionFromToolVersionsFile`. [[1]](diffhunk://#diff-69c9100f2a3041c22e8fddf77d13fd7eb1aa42a8a200aa35a81ea589c135e66aR77-R87) [[2]](diffhunk://#diff-69c9100f2a3041c22e8fddf77d13fd7eb1aa42a8a200aa35a81ea589c135e66aL85-R100)
* Implemented `resolveLatestPatchVersion` to probe the download host and resolve a major.minor version (or wildcard patch) to the latest available patch version, with robust error handling for missing releases and network errors.
* Updated the main `run` logic to support resolving major.minor and wildcard patch versions before proceeding with download.

**Testing improvements:**

* Added comprehensive tests for major.minor and wildcard patch version handling, including validation, patch resolution, network error propagation, and edge cases like missing releases or excessive matches. [[1]](diffhunk://#diff-857af24997d168282b0e700ede9dc3e987fbf72c6797e997bdf00f9ee6448b3bL212-R228) [[2]](diffhunk://#diff-857af24997d168282b0e700ede9dc3e987fbf72c6797e997bdf00f9ee6448b3bR241-R258) [[3]](diffhunk://#diff-857af24997d168282b0e700ede9dc3e987fbf72c6797e997bdf00f9ee6448b3bR323-R468)
* Switched from `vi.stubGlobal` to `vi.spyOn(globalThis, 'fetch')` in tests to ensure reliable restoration of mocks and prevent test leakage.

**Error messaging:**

* Improved error messages for invalid version entries in `.tool-versions`, clarifying the accepted formats and providing actionable guidance.

These changes improve flexibility for users specifying Helm versions and ensure robust, predictable behavior across a wider range of inputs.

This PR once merges will fix #109

### CODE REVIEW

File: src/run.ts [hunk 1]
```diff
@@ -82,7 +93,7 @@ export function getVersionFromToolVersionsFile(filePath: string): string {
    if (!version) {
       throw new Error(`No helm version found in '${filePath}'`)
    }
-   if (!isSemVerShaped(version)) {
+   if (!isSemVerShaped(version) && !isMajorMinorShaped(version)) {
       throw new Error(
          `The helm version '${version}' in '${filePath}' is not a valid semantic version`
       )
```
[the remark this answers is not in the document]
Comment (ejaifeobuks) [reply src/run.ts:103]: Will update in the next commit

File: src/run.ts [hunk 2]
```diff
@@ -121,6 +132,66 @@ export async function getLatestHelmVersion(): Promise<string> {
...
+// caller can distinguish an outage from a missing patch.
+export async function helmPatchExists(
+   baseURL: string,
+   version: string
+): Promise<boolean> {
+   const response = await fetch(getHelmDownloadURL(baseURL, version), {
+      method: 'HEAD'
+   })
+   return response.ok
+}
```
[the remark this answers is not in the document]
Comment (ejaifeobuks) [reply src/run.ts]: Will fix this in the next commit

File: src/run.test.ts [hunk 3]
```diff
@@ -296,6 +314,110 @@ describe('run.ts', () => {
...
+   // existing (200) and everything else as missing (404).
+   const stubPatchProbes = (existing: string[]) => {
+      const present = new Set(existing)
+      vi.stubGlobal(
+         'fetch',
+         vi.fn(async (url: string) => {
+            const version = url.match(/helm-(v\d+\.\d+\.\d+)-/)?.[1] ?? ''
+            return {ok: present.has(version)} as Response
+         })
+      )
```
[the remark this answers is not in the document]
Comment (ejaifeobuks) [reply src/run.test.ts]: This issue is now resolved in the latest commit

File: src/run.ts [hunk 4]
```diff
@@ -121,6 +134,75 @@ export async function getLatestHelmVersion(): Promise<string> {
...
+   if (!(await helmPatchExists(baseURL, `v${major}.${minor}.0`))) {
+      throw new Error(`No Helm releases found for ${major}.${minor}`)
+   }
+
+   let latestPatch = 0
+   let consecutiveMisses = 0
+   for (
+      let patch = 1;
+      patch <= maxPatch && consecutiveMisses < patchLookahead;
+      patch++
```
[the remark this answers is not in the document]
Comment (ejaifeobuks) [reply src/run.ts:241]: Valid concern. However, for helm this is not an issue since helm patch releases tend to be contiguous, so this is unlikely to happen

File: src/run.ts [hunk 5]
```diff
@@ -121,6 +134,75 @@ export async function getLatestHelmVersion(): Promise<string> {
...
+   baseURL: string,
+   version: string
+): Promise<string> {
+   const [major, minor] = (
+      version[0] === 'v' ? version.slice(1) : version
+   ).split('.')
+
+   if (!(await helmPatchExists(baseURL, `v${major}.${minor}.0`))) {
+      throw new Error(`No Helm releases found for ${major}.${minor}`)
+   }
```
[the remark this answers is not in the document]
Comment (ejaifeobuks) [reply src/run.ts:234]: Valid concern. However, Helm always publishes **X.Y.0** as the first patch for any minor so there is no minor that starts at .**1.** As such this precheck is sound

**Comment by davidgamero** [comment]
instead of interating through every patch version sequentially, consider a prefix listing 

ex:

```
get.helm.sh/?restype=container&comp=list&prefix=helm-v3.14.
// note the trailing period limits to minor 14 otherwise v3.140 would also match albeit unlikely to exist
// returns xml listing only blobs under that prefix. one request
```

**Comment by ejaifeobuks** [comment]
> instead of interating through every patch version sequentially, consider a prefix listing
> 
> ex:
> 
> ```
> get.helm.sh/?restype=container&comp=list&prefix=helm-v3.14.
> // note the trailing period limits to minor 14 otherwise v3.140 would also match albeit unlikely to exist
> // returns xml listing only blobs under that prefix. one request
> ```

Thanks for this David. I did consider this approach but I later had to abandon it. enumerating versions relies on Azure Blob Storage's specific listing API, which only works if the download host is Azure storage. A custom mirror (S3, Artifactory, nginx, etc.) usually doesn't support it — so it would break for mirror users

File: src/run.ts [hunk 6]
```diff
@@ -72,6 +74,17 @@ export function isSemVerShaped(version: string): boolean {
    return semVerShape.test(version)
 }
 
+// Matches a major.minor version with an optional leading 'v' and either no
+// patch component or a wildcard patch ('.x' / '.*'), e.g. '3.14', 'v3.14',
+// '3.14.x', 'v3.14.*'.
+const majorMinorShape = /^v?\d+\.\d+(?:\.[x*])?$/
```
Comment (davidgamero) [inline src/run.ts:82]: if we are going to use regex here, we should use the one from semver https://semver.org/#is-there-a-suggested-regular-expression-regex-to-check-a-semver-string

File: src/run.ts [hunk 7]
```diff
@@ -121,6 +134,75 @@ export async function getLatestHelmVersion(): Promise<string> {
...
+// the artifact exists (2xx) and false only when it is definitively absent
+// (404). Any other status (403/405/429/5xx, ...) and genuine network errors are
+// thrown, so transient failures, rate-limiting, or a host that disallows HEAD
+// are never mistaken for a missing patch.
+export async function helmPatchExists(
+   baseURL: string,
+   version: string
+): Promise<boolean> {
+   const url = getHelmDownloadURL(baseURL, version)
+   const response = await fetch(url, {method: 'HEAD'})
```
Comment (davidgamero) [inline src/run.ts:154]: do mirrors typically support HEAD method?

**Comment by davidgamero** [comment]
> > instead of interating through every patch version sequentially, consider a prefix listing
> > ex:
> > ```
> > get.helm.sh/?restype=container&comp=list&prefix=helm-v3.14.
> > // note the trailing period limits to minor 14 otherwise v3.140 would also match albeit unlikely to exist
> > // returns xml listing only blobs under that prefix. one request
> > ```
> 
> Thanks for this David. I did consider this approach but I later had to abandon it. enumerating versions relies on Azure Blob Storage's specific listing API, which only works if the download host is Azure storage. A custom mirror (S3, Artifactory, nginx, etc.) usually doesn't support it — so it would break for mirror users

i see. what about detecting if it's an azure blob store to then use this optimization?

File: src/run.ts (hunk 7 above)
Comment (ejaifeobuks) [reply src/run.ts:154]: Yes they do. HEAD is a universal standard method that comes for free with plain file serving which is the exact capability a mirror must support to serve downloads at all

File: src/run.ts (hunk 6 above)
Comment (ejaifeobuks) [reply src/run.ts:82]: Hi David, thanks for the feedback! In this case I don't think the semver.org regex fits, since it only validates complete X.Y.Z strings and has no notation for the partial versions this pattern is meant to match (e.g. 3.14, 3.14.x, 3.14.*). Swapping it in here would actually break the latest-patch feature, which depends on accepting an incomplete version string to resolve the newest patch.

That said, the semver.org regex is a great fit for our full version check, so I will adopt it for the  ```(isSemVerShape)```regex instead

**Comment by ejaifeobuks** [comment]
> i see. what about detecting if it's an azure blob store to then use this optimization?

Nice suggestion! However, I don't think upfront detection is worth the overhead (and hostname sniffing would actually miss get.helm.sh, since it's a CNAME over Azure Blob). Instead, I've added the listing as an optimization with a fallback: the action first tries to enumerate from whatever download host the binaries live at, and if that returns no valid listing it falls back to the iterative probe. Same-host enumeration means we never resolve a version the host can't serve, and it works on any mirror.

**Review by davidgamero — APPROVED** [review]
lgtm
